File: src/rl_fzerox/apps/run_manager/api/handlers/save_games.py
# src/rl_fzerox/apps/run_manager/api/handlers/save_games.py
from __future__ import annotations

from pathlib import Path

from fastapi import HTTPException

from rl_fzerox.apps.run_manager.api.contracts import (
    CreateSaveGameRequest,
    ImportSaveEngineTuningRequest,
    UpdateSaveGameRequest,
    UpdateSaveRunnerSettingsRequest,
    UpsertSaveCourseSetupRequest,
    UpsertSaveCupSetupRequest,
)
from rl_fzerox.apps.run_manager.api.handlers.save_game_status import (
    save_game_payload_for_store,
    save_game_status_payload_for_store,
)
from rl_fzerox.apps.run_manager.desktop import open_directory
from rl_fzerox.core.engine_tuning import (
    EngineTuningContext,
    OrderedEngineTuner,
)
from rl_fzerox.core.engine_tuning.config import engine_tuner_settings
from rl_fzerox.core.manager import (
    ManagedPolicySource,
    ManagedRun,
    ManagerStore,
    PolicySourceArtifact,
)
from rl_fzerox.core.manager.errors import ManagerNameConflictError
from rl_fzerox.core.manager.projection.engine_tuning import adaptive_engine_tuning_config
from rl_fzerox.core.training.runs import resolve_policy_artifact_path
from rl_fzerox.core.training.session.artifacts import load_engine_tuning_checkpoint_state


def save_games_payload(store: ManagerStore) -> dict[str, list[dict[str, object]]]:
    items = store.list_save_games()
    return {"save_games": [save_game_payload_for_store(store, item) for item in items]}


def save_game_status_payload_for_id(
    store: ManagerStore,
    save_game_id: str,
) -> dict[str, dict[str, object]]:
    save_game = store.get_save_game(save_game_id)
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"save_game": save_game_status_payload_for_store(store, save_game)}


def create_save_game_payload(
    store: ManagerStore,
    request: CreateSaveGameRequest,
    name: str,
) -> dict[str, dict[str, object]]:
    try:
        del request
        save_game = store.create_save_game(name=name)
    except ManagerNameConflictError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"save_game": save_game_payload_for_store(store, save_game)}


def update_save_game_payload(
    store: ManagerStore,
    save_game_id: str,
    request: UpdateSaveGameRequest,
) -> dict[str, dict[str, object]]:
    try:
        save_game = store.rename_save_game(save_game_id=save_game_id, name=request.name)
    except ManagerNameConflictError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"save_game": save_game_payload_for_store(store, save_game)}


def update_save_game_runner_settings_payload(
    store: ManagerStore,
    save_game_id: str,
    request: UpdateSaveRunnerSettingsRequest,
) -> dict[str, dict[str, object]]:
    try:
        save_game = store.update_save_game_runner_settings(
            save_game_id=save_game_id,
            device=request.device,
            renderer=request.renderer,
            policy_mode=request.policy_mode,
            attempt_seed=request.attempt_seed,
            recording_enabled=request.recording_enabled,
            recording_input_hud_enabled=request.recording_input_hud_enabled,
            recording_upscale_factor=request.recording_upscale_factor,
            recording_path=request.recording_path,
            target_restart_on_retire=request.target_restart_on_retire,
            target_clear_goal=request.target_clear_goal,
            keep_failed_recordings=request.keep_failed_recordings,
            reload_policy_between_attempts=request.reload_policy_between_attempts,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"save_game": save_game_payload_for_store(store, save_game)}


def delete_save_game_payload(store: ManagerStore, save_game_id: str) -> dict[str, bool]:
    try:
        deleted = store.delete_save_game(save_game_id)
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if not deleted:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"deleted": True}


def upsert_save_course_setup_payload(
    store: ManagerStore,
    save_game_id: str,
    request: UpsertSaveCourseSetupRequest,
) -> dict[str, dict[str, object]]:
    try:
        store.upsert_save_course_setup(
            save_game_id=save_game_id,
            difficulty=request.difficulty,
            cup_id=request.cup_id,
            course_id=request.course_id,
            policy_source_kind=request.policy_source_kind,
            policy_source_id=request.policy_source_id,
            policy_artifact=request.policy_artifact,
            engine_setting_raw_value=request.engine_setting_raw_value,
        )
    except KeyError as error:
        raise HTTPException(status_code=404, detail=str(error).strip("'")) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    save_game = store.get_save_game(save_game_id)
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"save_game": save_game_payload_for_store(store, save_game)}


def upsert_save_cup_setup_payload(
    store: ManagerStore,
    save_game_id: str,
    request: UpsertSaveCupSetupRequest,
) -> dict[str, dict[str, object]]:
    try:
        store.upsert_save_cup_setup(
            save_game_id=save_game_id,
            difficulty=request.difficulty,
            cup_id=request.cup_id,
            vehicle_id=request.vehicle_id,
        )
    except KeyError as error:
        raise HTTPException(status_code=404, detail=str(error).strip("'")) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    save_game = store.get_save_game(save_game_id)
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    return {"save_game": save_game_payload_for_store(store, save_game)}


def import_save_engine_tuning_payload(
    store: ManagerStore,
    save_game_id: str,
    request: ImportSaveEngineTuningRequest,
) -> dict[str, object]:
    save_game = store.get_save_game(save_game_id)
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    try:
        policy_source = store.resolve_policy_source(
            policy_source_kind=request.policy_source_kind,
            policy_source_id=request.policy_source_id,
            policy_artifact=request.policy_artifact,
            require_policy_artifact=request.policy_source_kind == "evaluation",
        )
    except KeyError as error:
        raise HTTPException(status_code=404, detail=str(error).strip("'")) from error
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    if policy_source.config.vehicle.engine_mode != "adaptive_tuner":
        raise HTTPException(status_code=400, detail="policy source has no adaptive engine tuning")
    try:
        policy_path = _resolve_engine_tuning_policy_path(store, policy_source)
    except FileNotFoundError as error:
        raise HTTPException(
            status_code=400,
            detail=f"policy source has no {request.policy_artifact} artifact",
        ) from error
    state = load_engine_tuning_checkpoint_state(policy_path)
    if state is None or (not state.candidates and state.model_state is None):
        raise HTTPException(status_code=400, detail="policy source has no engine tuning samples")
    if not request.course_setups:
        raise HTTPException(
            status_code=400,
            detail="no course setup drafts use this policy source",
        )
    tuner = OrderedEngineTuner(
        settings=engine_tuner_settings(adaptive_engine_tuning_config(policy_source.config)),
        state=state,
    )
    recommendations: list[dict[str, object]] = []
    for setup in request.course_setups:
        recommendation = tuner.recommendation(
            EngineTuningContext(
                course_key=setup.course_id,
                vehicle_id=setup.vehicle_id,
            )
        )
        recommendations.append(
            {
                "difficulty": setup.difficulty,
                "cup_id": setup.cup_id,
                "course_id": setup.course_id,
                "vehicle_id": setup.vehicle_id,
                "engine_setting_raw_value": recommendation.engine_setting_raw_value,
                "mean_score": recommendation.mean_score,
                "finish_count": recommendation.finish_count,
            }
        )
    return {"recommendations": recommendations}


def open_save_game_dir_payload(store: ManagerStore, save_game_id: str) -> dict[str, bool]:
    save_game = store.get_save_game(save_game_id)
    if save_game is None:
        raise HTTPException(status_code=404, detail="save game not found")
    try:
        open_directory(save_game.save_path.parent)
    except RuntimeError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return {"opened": True}


def _resolve_engine_tuning_policy_path(
    store: ManagerStore,
    policy_source: ManagedPolicySource,
) -> Path:
    if policy_source.kind == "evaluation":
        if policy_source.policy_path is None:
            raise FileNotFoundError("evaluation policy checkpoint is missing")
        return policy_source.policy_path
    run = store.get_run(policy_source.id)
    if run is None:
        raise FileNotFoundError(f"policy run not found: {policy_source.id}")
    return _resolve_engine_tuning_run_policy_artifact(run, artifact=policy_source.artifact)


def _resolve_engine_tuning_run_policy_artifact(
    run: ManagedRun,
    *,
    artifact: PolicySourceArtifact,
) -> Path:
    try:
        return resolve_policy_artifact_path(run.run_dir, artifact=artifact)
    except FileNotFoundError:
        if run.source_snapshot_dir is None:
            raise
    return resolve_policy_artifact_path(run.source_snapshot_dir, artifact=artifact)

File: web/run-manager/src/shared/api/client/resources/saveGames.ts
// web/run-manager/src/shared/api/client/resources/saveGames.ts
import { getJson, parseApiPayload, parseJson } from "@/shared/api/client/http";
import {
  type CareerModeRunnerLaunchRequest,
  createSaveGameResponseSchema,
  deleteSaveGameResponseSchema,
  importSaveEngineTuningResponseSchema,
  type ManagedSaveGame,
  type ManagedSaveGameStatus,
  openSaveGameDirectoryResponseSchema,
  type SaveEngineTuningCourseSetupRecommendation,
  type SaveGameRunnerSettingsUpdateRequest,
  type SavePolicyArtifact,
  saveGameStatusResponseSchema,
  saveGamesResponseSchema,
  updateSaveGameRunnerSettingsResponseSchema,
  upsertSaveCourseSetupResponseSchema,
  watchRunResponseSchema,
} from "@/shared/api/contract";

export async function fetchSaveGames(): Promise<ManagedSaveGame[]> {
  const payload = parseApiPayload(saveGamesResponseSchema, await getJson("/api/save-games"));
  return payload.save_games;
}

export async function fetchSaveGameStatus(saveGameId: string): Promise<ManagedSaveGameStatus> {
  const payload = parseApiPayload(
    saveGameStatusResponseSchema,
    await getJson(`/api/save-games/${encodeURIComponent(saveGameId)}/status`),
  );
  return payload.save_game;
}

export async function createSaveGame(name: string): Promise<ManagedSaveGame> {
  const response = await fetch("/api/save-games", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
  const payload = parseApiPayload(createSaveGameResponseSchema, await parseJson(response));
  return payload.save_game;
}

export async function renameSaveGame(saveGameId: string, name: string): Promise<ManagedSaveGame> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
  const payload = parseApiPayload(createSaveGameResponseSchema, await parseJson(response));
  return payload.save_game;
}

export async function updateSaveGameRunnerSettings({
  attemptSeed,
  device,
  policyMode,
  recordingEnabled,
  recordingInputHudEnabled,
  recordingUpscaleFactor,
  recordingPath,
  renderer,
  saveGameId,
  targetRestartOnRetire,
  targetClearGoal,
  keepFailedRecordings,
  reloadPolicyBetweenAttempts,
}: SaveGameRunnerSettingsUpdateRequest): Promise<ManagedSaveGame> {
  const response = await fetch(
    `/api/save-games/${encodeURIComponent(saveGameId)}/runner-settings`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        device,
        renderer,
        attempt_seed: attemptSeed === null ? null : Number(attemptSeed),
        policy_mode: policyMode,
        recording_enabled: recordingEnabled,
        recording_input_hud_enabled: recordingInputHudEnabled,
        recording_upscale_factor: recordingUpscaleFactor,
        recording_path: recordingPath,
        target_restart_on_retire: targetRestartOnRetire,
        target_clear_goal: targetClearGoal,
        keep_failed_recordings: keepFailedRecordings,
        reload_policy_between_attempts: reloadPolicyBetweenAttempts,
      }),
    },
  );
  const payload = parseApiPayload(
    updateSaveGameRunnerSettingsResponseSchema,
    await parseJson(response),
  );
  return payload.save_game;
}

export async function deleteSaveGame(saveGameId: string): Promise<void> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}`, {
    method: "DELETE",
  });
  parseApiPayload(deleteSaveGameResponseSchema, await parseJson(response));
}

export async function openSaveGameDirectory(saveGameId: string): Promise<void> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}/open-dir`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
  });
  parseApiPayload(openSaveGameDirectoryResponseSchema, await parseJson(response));
}

export async function startCareerModeRunner({
  attemptSeed,
  device,
  policyMode,
  recordingEnabled,
  recordingInputHudEnabled,
  recordingUpscaleFactor,
  recordingPath,
  renderer,
  saveGameId,
  singleTarget,
  perfectRun,
  keepFailedRecordings,
  reloadPolicyBetweenAttempts,
  targetClearGoal,
  target,
}: CareerModeRunnerLaunchRequest): Promise<"started" | "already_running"> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}/runner`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      device,
      renderer,
      attempt_seed: attemptSeed === null ? null : Number(attemptSeed),
      policy_mode: policyMode,
      recording_enabled: recordingEnabled,
      recording_input_hud_enabled: recordingInputHudEnabled,
      recording_upscale_factor: recordingUpscaleFactor,
      recording_path: recordingEnabled ? recordingPath : null,
      single_target: singleTarget,
      perfect_run: perfectRun,
      keep_failed_recordings: keepFailedRecordings,
      reload_policy_between_attempts: reloadPolicyBetweenAttempts,
      target_clear_goal: targetClearGoal,
      ...(target === null
        ? {}
        : {
            target_kind: target.kind,
            difficulty: target.difficulty,
            cup_id: target.cup_id,
            course_id: target.course_id,
          }),
    }),
  });
  const payload = parseApiPayload(watchRunResponseSchema, await parseJson(response));
  return payload.status;
}

export async function upsertSaveCourseSetup({
  courseId,
  cupId,
  difficulty,
  engineSettingRawValue,
  policyArtifact,
  policyRunId,
  saveGameId,
}: {
  courseId?: string | null;
  cupId?: string | null;
  difficulty?: string | null;
  engineSettingRawValue: number;
  policyArtifact: SavePolicyArtifact;
  policyRunId: string;
  saveGameId: string;
}): Promise<ManagedSaveGame> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}/course-setups`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      difficulty: difficulty ?? null,
      cup_id: cupId ?? null,
      course_id: courseId ?? null,
      policy_run_id: policyRunId,
      policy_artifact: policyArtifact,
      engine_setting_raw_value: engineSettingRawValue,
    }),
  });
  const payload = parseApiPayload(upsertSaveCourseSetupResponseSchema, await parseJson(response));
  return payload.save_game;
}

export async function importSaveEngineTuning({
  courseSetups,
  policyArtifact,
  policyRunId,
  saveGameId,
}: {
  courseSetups: readonly {
    courseId: string;
    cupId: string;
    difficulty?: string | null;
    vehicleId: string;
  }[];
  policyArtifact: SavePolicyArtifact;
  policyRunId: string;
  saveGameId: string;
}): Promise<readonly SaveEngineTuningCourseSetupRecommendation[]> {
  const response = await fetch(
    `/api/save-games/${encodeURIComponent(saveGameId)}/course-setups/import-engine-tuning`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        course_setups: courseSetups.map((setup) => ({
          difficulty: setup.difficulty ?? null,
          cup_id: setup.cupId,
          course_id: setup.courseId,
          vehicle_id: setup.vehicleId,
        })),
        policy_artifact: policyArtifact,
        policy_run_id: policyRunId,
      }),
    },
  );
  const payload = parseApiPayload(importSaveEngineTuningResponseSchema, await parseJson(response));
  return payload.recommendations;
}

export async function upsertSaveCupSetup({
  cupId,
  difficulty,
  saveGameId,
  vehicleId,
}: {
  cupId: string;
  difficulty?: string | null;
  saveGameId: string;
  vehicleId: string;
}): Promise<ManagedSaveGame> {
  const response = await fetch(`/api/save-games/${encodeURIComponent(saveGameId)}/cup-setups`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cup_id: cupId,
      difficulty: difficulty ?? null,
      vehicle_id: vehicleId,
    }),
  });
  const payload = parseApiPayload(upsertSaveCourseSetupResponseSchema, await parseJson(response));
  return payload.save_game;
}
